### PR TITLE
Use more flexible implementation for user mention link documentation extension

### LINKED ISSUE #3542: Ignore `@` in code blocks when generating GitHub username links
## Description

In https://github.com/WordPress/openverse/pull/3523 @AetherUnbound pointed out that the link usernames Sphinx plugin is transforming the `@message` notation inside the code block to a GitHub user link.

We should ignore the code block context in that Sphinx plugin.

## Additional context

I'm looking at the code now for the link usernames Sphinx plugin and see that it's just a naive regex. That's not great for flexibility in cases like this :cry:

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Fixes #3542 by @sarayourfriend

## Description

This PR changes the implementation of the `link_usernames` extension to be more flexible and resilient to the various text situations that can exist.

In doing so, it now trivially ignores the following contexts:
* Links
* Back `ticks`
* Code blocks (```` ``` ````)

The previous implementation could not ignore any of these without roundabout methods, like requiring leading whitespace. This led to the use of the flaky list of usernames to ignore. The new implementation doesn't need such a list, as it handles all the cases the ignore list targeted without issue, and removes the otherwise arbitrary requirement for leading whitespace. This means forms like `@person-a/@person-b` will work perfectly.

Additionally, the new implementation handles team mentions just as easily as username mentions.

The implementation is a significantly simplified version of the issue reference.

## Testing Instructions

Review the code and confirm it makes sense. Check out the documentation preview on this PR and confirm that usernames are handled correctly and that the known edge cases are still handled, in addition to the new fix for code blocks (note, this is proven to be fixed by the removal of the `todo` and `defaultValue` ignored usernames. Those only appeared in code blocks!).

## Checklist

- [x] My pull request has a descriptive title (not a vague title like`Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [N/A] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.
- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: documentation/_ext/link_usernames.py [hunk 1]
```diff
@@ -4,47 +4,115 @@
...
+# Format to use for team mentions
+GITHUB_TEAM_TEMPLATE = "https://github.com/orgs/{org}/teams/{team}"
+
 # Regex to match username references
 GITHUB_USERNAME_REGEX = re.compile(
     r"""
-\ # Required initial whitespace
-@([A-Za-z0-9-]+)  # Match @ then any alphanumeric character or - for the username
+@(?P<username>[A-Za-z0-9-]+)  # Match @ then any alphanumeric character or - for the username
+(?P<team>/[A-Za-z0-9-]+)?  # If this is a team reference, match the team separately
```
Comment (dhruvkb) [inline documentation/_ext/link_usernames.py:30]: At the risk of slightly complicating the regex, we can keep the leading `/` out of the team group. Also `A-Za-z0-9` can be simplified as `\w`.

```suggestion
@(?P<username>[\w-]+)  # Match @ then any alphanumeric character or - for the username
(/(?P<team>[\w-]+))?  # If this is a team reference, match the team separately
```

Also TIL comments work inside triple-quoted strings!

**Review by dhruvkb — APPROVED** [review]
This works well. I tried it with all possible combinations of usernames and team names, both outside and inside code blocks, both inline and standalone.

This works well in all cases.

File: documentation/_ext/link_usernames.py (hunk 1 above)
Comment (sarayourfriend) [reply documentation/_ext/link_usernames.py:30]: Yeah! But only with the verbose flag, mind: https://docs.python.org/3/library/re.html?highlight=re%20verbose#re.VERBOSE

**Review by AetherUnbound — APPROVED** [review]
This is excellent, thanks for making it more robust! I love that we have better titles now and that it handles teams references too 🚀

**Pull request merged**